### PR TITLE
Add shortlink updater interface to use cases

### PR DESCRIPTION
**Pull request by oatovar**
### Description
* Connect shortlink update usecase with sql adapter.
* Add shortlink updater interface
* Implement shortlink updater interface 
* Add unit tests for url mutation edge cases.

### CODE REVIEW

File: backend/app/usecase/shortlink/updater.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	"github.com/short-d/short/backend/app/entity"
+	"github.com/short-d/short/backend/app/usecase/keygen"
+	"github.com/short-d/short/backend/app/usecase/repository"
+	"github.com/short-d/short/backend/app/usecase/risk"
+	"github.com/short-d/short/backend/app/usecase/validator"
+)
+
+var _ Updater = (*UpdaterPersist)(nil)
+
+// ErrShortLinkNotFound means that a queried shortlink was not found
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:16]:
```suggestion
// ErrShortLinkNotFound represents the failure of finding certain short link in the data store
```

File: backend/app/usecase/shortlink/updater.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+var ErrShortLinkNotFound = errors.New("short link not found")
+
+// Updater modifies the properties of existing short links.
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:19]:
```suggestion
// Updater mutates existing short links.
```

File: backend/app/usecase/shortlink/updater.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+type Updater interface {
+	UpdateShortLink(oldAlias string, update entity.ShortLink, user entity.User) (entity.ShortLink, error)
+}
+
+// UpdaterPersist persists the given changes to a short link in the database store.
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:24]:
```suggestion
// UpdaterPersist persists the mutated short link in the data store..
```

File: backend/app/usecase/shortlink/updater.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+type UpdaterPersist struct {
+	shortLinkRepo             repository.ShortLink
+	userShortLinkRelationRepo repository.UserShortLink
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:27]:
```suggestion
	userShortLinkRepo repository.UserShortLink
```

File: backend/app/usecase/shortlink/updater.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	keyGen                    keygen.KeyGenerator
+	longLinkValidator         validator.LongLink
+	aliasValidator            validator.CustomAlias
+	timer                     timer.Timer
+	riskDetector              risk.Detector
+}
+
+// UpdateShortLink persists mutations for a given short link in the repository.
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:35]:
```suggestion
// UpdateShortLink mutates a short link in the repository.
```

File: backend/app/usecase/shortlink/updater.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+func (u UpdaterPersist) UpdateShortLink(
+	oldAlias string,
+	update entity.ShortLink,
+	user entity.User,
+) (entity.ShortLink, error) {
+	ok, err := u.userShortLinkRelationRepo.HasMapping(user, oldAlias)
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:41]:
```suggestion
	hasMapping, err := u.userShortLinkRelationRepo.HasMapping(user, oldAlias)
```

File: backend/app/usecase/shortlink/updater.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	if err != nil {
+		return entity.ShortLink{}, err
+	}
+
+	shortLink = u.updateAlias(shortLink, update)
+	shortLink = u.updateLongLink(shortLink, update)
+
+	if isValid, violation := u.aliasValidator.IsValid(&shortLink.Alias); !isValid {
+		return entity.ShortLink{}, ErrInvalidCustomAlias{shortLink.Alias, violation}
+	}
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:59]: Prefer separating them into two lines so that it's easier to follow.

File: backend/app/usecase/shortlink/updater.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+	if isValid, violation := u.longLinkValidator.IsValid(&shortLink.LongLink); !isValid {
+		return entity.ShortLink{}, ErrInvalidLongLink{shortLink.LongLink, violation}
+	}
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:63]: Ditto.

File: backend/app/usecase/shortlink/updater.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+	return shortLink
+}
+
+// NewUpdaterPersist creates a new UpdaterPersist instance.
+func NewUpdaterPersist(
+	shortLinkRepo repository.ShortLink,
+	userShortLinkRelationRepo repository.UserShortLink,
+	keyGen keygen.KeyGenerator,
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:97]: Are we actually using it?

File: backend/app/usecase/shortlink/updater.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+		shortLink.LongLink = newLongLink
+	}
+
+	return shortLink
+}
+
+// NewUpdaterPersist creates a new UpdaterPersist instance.
+func NewUpdaterPersist(
+	shortLinkRepo repository.ShortLink,
+	userShortLinkRelationRepo repository.UserShortLink,
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:96]:
```suggestion
	userShortLinkRepo repository.UserShortLink,
```

File: backend/app/usecase/shortlink/updater_test.go [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+	"github.com/short-d/short/backend/app/usecase/validator"
+)
+
+func TestShortLinkUpdaterPersist_UpdateShortLink(t *testing.T) {
+	t.Parallel()
+
+	now := time.Now().UTC()
+	alias := "boGp9w35"
+	validNewAlias := "eBJRJJty"
+	validNewLongLink := "https://httpbin.org/get?p1=v1"
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater_test.go:24]: Let's embed them into the test case directly  so that we can update the tests independently in the future.

Comment (byliuyang) [reply backend/app/usecase/shortlink/updater_test.go:24]: https://github.com/short-d/short#test-structure

File: backend/app/usecase/shortlink/updater_test.go [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+	testCases := []struct {
+		name               string
+		alias              string
+		availableKeys      []keygen.Key
+		shortlinks         shortLinks
+		user               entity.User
+		shortLinkUpdate    entity.ShortLink
+		relationUsers      []entity.User
+		relationShortLinks []entity.ShortLink
+		expHasErr          bool
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater_test.go:35]:
```suggestion
		expectedHasErr          bool
```

File: backend/app/usecase/shortlink/updater_test.go [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+
+	now := time.Now().UTC()
+	alias := "boGp9w35"
+	validNewAlias := "eBJRJJty"
+	validNewLongLink := "https://httpbin.org/get?p1=v1"
+
+	testCases := []struct {
+		name               string
+		alias              string
+		availableKeys      []keygen.Key
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater_test.go:29]: This is unused in the codebase.

File: backend/app/usecase/shortlink/updater_test.go [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+		shortlinks         shortLinks
+		user               entity.User
+		shortLinkUpdate    entity.ShortLink
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater_test.go:32]:
```suggestion
		update    entity.ShortLink
```

File: backend/app/usecase/shortlink/updater_test.go [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+					Alias:     "boGp9w35",
+					LongLink:  "https://httpbin.org",
+					UpdatedAt: &now,
+				},
+			},
+			expHasErr:         true,
+			expectedShortLink: entity.ShortLink{},
+		},
+		{
+			name:  "malicious shortLink update",
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater_test.go:162]:
```suggestion
			name:  "reject malicious long link",
```

File: backend/app/usecase/shortlink/updater_test.go [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+			expHasErr:         true,
+			expectedShortLink: entity.ShortLink{},
+		},
+	}
+
+	for _, testCase := range testCases {
+		testCase := testCase
+		t.Run(testCase.name, func(t *testing.T) {
+			t.Parallel()
+			blockedHash := map[string]bool{
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater_test.go:179]:
```suggestion
			blockedHash := map[string]struct{}{
```
Use empty struct to save some spaces.

File: backend/app/usecase/shortlink/updater_test.go [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+				"http://malware.wicar.org/data/ms14_064_ole_not_xp.html": false,
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater_test.go:180]: Let's make this part of test data so that it helps reader to understand the test case when going through the table.

File: backend/app/usecase/shortlink/updater_test.go [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+			}
+			tm := timer.NewStub(now)
+			shortLinkRepo := repository.NewShortLinkFake(testCase.shortlinks)
+			userShortLinkRepo := repository.NewUserShortLinkRepoFake(
+				testCase.relationUsers,
+				testCase.relationShortLinks,
+			)
+			keyFetcher := keygen.NewKeyFetcherFake(testCase.availableKeys)
+			keyGen, err := keygen.NewKeyGenerator(2, &keyFetcher)
+			assert.Equal(t, nil, err)
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater_test.go:190]: Those are not used in the updater.

**Comment by byliuyang** [comment]
Hi @oatovar , thank you so much for working on this feature! I left few comment on the PR!

File: backend/app/usecase/shortlink/updater.go (hunk 1 above)
Comment (oatovar) [reply backend/app/usecase/shortlink/updater.go:16]: Update ErrShortLinkNotFound go doc

File: backend/app/usecase/shortlink/updater.go (hunk 2 above)
Comment (oatovar) [reply backend/app/usecase/shortlink/updater.go:19]: Cleanup Updater go doc

File: backend/app/usecase/shortlink/updater.go (hunk 9 above)
Comment (oatovar) [reply backend/app/usecase/shortlink/updater.go:97]: I'll remove it :smiley:

File: backend/app/usecase/shortlink/updater_test.go (hunk 16 above)
Comment (oatovar) [reply backend/app/usecase/shortlink/updater_test.go:179]: It was originally changed to bool in a previous code review to avoid the weird `map[string]struct{}{}` syntax. Changing the map value type will force a lot of tests to update. I don't think the byte difference is worth changing it at this point. Maybe in the future?

Comment (byliuyang) [reply backend/app/usecase/shortlink/updater_test.go]: @oatovar Sounds good to me!

File: backend/app/usecase/shortlink/updater.go [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+
+// ErrShortLinkNotFound represents the failure of finding certain short link in the data store
+var ErrShortLinkNotFound = errors.New("short link not found")
+
+// Updater mutates existing short links.
+type Updater interface {
+	UpdateShortLink(oldAlias string, update entity.ShortLink, user entity.User) (entity.ShortLink, error)
+}
+
+// UpdaterPersist persists the mutated short link in the data store..
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:23]:
```suggestion
// UpdaterPersist persists the mutated short link in the data store.
```

File: backend/app/usecase/shortlink/updater.go [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+
+	return u.shortLinkRepo.UpdateShortLink(oldAlias, shortLink)
+}
+
+func (u UpdaterPersist) updateAlias(shortLink, update entity.ShortLink) entity.ShortLink {
+	newAlias := update.Alias
+	if newAlias != "" {
+		shortLink.Alias = newAlias
+	}
+
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:80]:
```suggestion
```

File: backend/app/usecase/shortlink/updater.go [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+	return shortLink
+}
+
+func (u *UpdaterPersist) updateLongLink(shortLink, update entity.ShortLink) entity.ShortLink {
+	newLongLink := update.LongLink
+	if newLongLink != "" {
+		shortLink.LongLink = newLongLink
+	}
+
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater.go:89]:
```suggestion
```

File: backend/app/usecase/shortlink/updater_test.go [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+			name:  "reject malicious long link",
+			alias: "boGp9w35",
+			user: entity.User{
+				Email: "gopher@golang.org",
+			},
+			update: entity.ShortLink{
+				LongLink: "http://malware.wicar.org/data/ms14_064_ole_not_xp.html",
+			},
+			blockedLongLinks: map[string]bool{
+				"http://malware.wicar.org/data/ms14_064_ole_not_xp.html": false,
```
Comment (byliuyang) [inline backend/app/usecase/shortlink/updater_test.go:167]:
```suggestion
				"http://malware.wicar.org/data/ms14_064_ole_not_xp.html": true,
```
